### PR TITLE
runtime-sdk: Restrict queries in secure builds

### PR DESCRIPTION
**Pull request by kostko**

### CODE REVIEW

File: runtime-sdk/src/keymanager.rs [hunk 1]
```diff
@@ -146,15 +170,6 @@ impl<'a> KeyManagerClientWithContext<'a> {
     }
 }
 
-impl<'a> Clone for KeyManagerClientWithContext<'a> {
```
Comment (pro-wh) [inline runtime-sdk/src/keymanager.rs:149]: :information_source: this is now derived

File: runtime-sdk/src/dispatcher.rs [hunk 2]
```diff
@@ -365,8 +366,19 @@ impl<R: Runtime> Dispatcher<R> {
...
-                _ => R::Modules::dispatch_query(ctx, method, args)
-                    .ok_or_else(|| modules::core::Error::InvalidMethod(method.into()))?,
+                _ => {
+                    // Even though queries are not allowed to perform any private key ops, we
+                    // reduce the attack surface by completely disallowing them in secure builds.
+                    //
+                    // If one needs to perform runtime queries, she should use a regular build of
+                    // the runtime for that.
+                    if BUILD_INFO.is_secure {
+                        return Err(modules::core::Error::InvalidMethod(method.into()).into());
```
Comment (pro-wh) [inline runtime-sdk/src/dispatcher.rs:376]: might be nice to have a separate error message for this condition. but I guess it's not critical for this

**Review by pro-wh — COMMENTED** [review]
was there any discussion around this? I see that this PR branch says it's a fix.

**Comment by kostko** [comment]
> was there any discussion around this? I see that this PR branch says it's a fix.

The discussion can happen in the PR. This is reducing the attack surface of secure builds that need to protect confidentiality as otherwise careless runtime modules could potentially leak private stuff via queries. Thoughts?

File: runtime-sdk/src/dispatcher.rs (hunk 2 above)
Comment (kostko) [reply runtime-sdk/src/dispatcher.rs:376]: Agreed, let's add a separate variant.

File: runtime-sdk/src/keymanager.rs (hunk 1 above)
Comment (kostko) [reply runtime-sdk/src/keymanager.rs:149]: Yeah not sure why it wasn't before.

**Comment by pro-wh** [comment]
I imagine an alternative to allowing queries in secure builds is that the runtime could make an unencrypted area of storage computed from private values during actual transactions. Then service queries from there. There's no quota or authentication for queries, so putting out the raw data would be pretty close.

I wonder if there are use cases where you'd want to enforce certain queries being run on confidential data. Oasis Labs is doing some work on differential privacy, but that's not an exact fit for queries, because there's also a privacy budget update involved. Maybe in these cases it'll be possible to precompute answers to some queries and store those instead of raw data.

**Review by pro-wh — APPROVED** [review]
alright

we can think about reenabling it if a desire comes up later

**Pull request merged**